Backend APIs

Adan-Asim-ASE opened the pull request.

Adan-Asim-ASE committed:
Backend Development code

Adan-Asim-ASE committed:
Removed node modules

Adan-Asim-ASE committed:
Modifcations in login api

kamranarshad1 commented on App.js:
```diff
@@ -0,0 +1,39 @@
+// }); 
+
+
+app.post("/CreateUser", controller.createUser);
```
`post` is self explanatory that we are creating user.  The end point just be `users`

kamranarshad1 commented on App.js:
```diff
@@ -0,0 +1,39 @@
+
+
+app.post("/CreateUser", controller.createUser);
+app.get("/CheckUser/:email", controller.findUser);
```
should be `/users/:email`

kamranarshad1 commented on App.js:
```diff
@@ -0,0 +1,39 @@
+app.get("/", (req, res) => {
+    res.json({ message: "Welcome to bezkoder application." });
+});
+// db.sequelize.sync({ force: true }).then(() => {
+//     console.log( "Drop and re-sync db.");
+// }); 
```
Please remove comments

kamranarshad1 commented on App.js:
```diff
@@ -0,0 +1,39 @@
+app.get("/CheckUser/:email", controller.findUser);
+
+app.post("/CreatePost/:id", controller.createPost);
+app.get("/AllPostsByUser/:id", controller.findAllPostsByUserId);
```
should be `posts/me`

kamranarshad1 commented on App.js:
```diff
@@ -0,0 +1,39 @@
+app.post("/CreateUser", controller.createUser);
+app.get("/CheckUser/:email", controller.findUser);
+
+app.post("/CreatePost/:id", controller.createPost);
```
should be `posts`. Please remove id and use current user's id from session

kamranarshad1 commented on config/db.config.js:
```diff
@@ -0,0 +1,7 @@
+    PASSWORD: "123",
+    DB: "project3",
+    dialect: "postgres",
+  };
```
Please load them them via env variables

kamranarshad1 commented on models/index.js:
```diff
@@ -0,0 +1,22 @@
+db.users = require("./users.model.js")(sequelize, Sequelize);
+db.posts = require("./posts.model.js")(sequelize, Sequelize);
+
+db.users.hasMany(db.posts, { as: "posts"});
+db.posts.belongsTo(db.users, {
+  foreignKey: "userId",
+  as: "user",
+})
```
These should be in users and posts models

kamranarshad1 commented on models/posts.model.js:
```diff
@@ -0,0 +1,14 @@
+module.exports = (sequelize, Sequelize) => {
+  const posts = sequelize.define("posts", {
+    title: {
+      type: Sequelize.STRING
+    },
+    content: {
+      type: Sequelize.STRING
+    },
+    published: {
+      type: Sequelize.BOOLEAN
+    }
```
model validations missing

kamranarshad1 commented on models/users.model.js:
```diff
@@ -0,0 +1,14 @@
+module.exports = (sequelize, Sequelize) => {
+  const users = sequelize.define("users", {
+    name: {
+      type: Sequelize.STRING
+    },
+    email: {
+      type: Sequelize.STRING
+    },
+    password: {
+      type: Sequelize.STRING
+    }
```
model validations missing

kamranarshad1 commented:
DB migrations are missing. Please create tables using db migrations with proper db constraints

Adan-Asim-ASE committed:
Modified as per code review

kamranarshad1 commented on .env:
```diff
@@ -0,0 +1,4 @@
+DB_USER=postgres
```
Please add this file to `.gitignore`

kamranarshad1 commented on App.js:
```diff
@@ -0,0 +1,29 @@
+
+require('dotenv').config();
+
+// db.sequelize.sync({ force: true }).then(() => {          
+//     console.log( "Drop and re-sync db.");
+// }); 
```
These have not been removed yet

Adan-Asim-ASE replied:
Actually need to use this code for recreating tables after modifications, therefore, didn't remove it, else have to search again and again.

kamranarshad1 commented on App.js:
```diff
@@ -0,0 +1,29 @@
+app.use(express.json());
+app.use(express.urlencoded({ extended: true }));
+
+app.use('/users', usersRoutes);
+app.use('/posts', postsRoutes);
```
Please create separate file to all routes and include it here

Adan-Asim-ASE replied:
Actually created two separate files for users routes and posts routes, all their related routes are there, except these base routes.

kamranarshad1 commented on controllers/posts.controller.js:
```diff
@@ -0,0 +1,125 @@
+    title: req.body.title,
+    content: req.body.content,
+    published: req.body.published,
+    userId: req.body.userId,
```
Still not using user from session

Adan-Asim-ASE committed:
Updations according to review

Adan-Asim-ASE committed:
Updations

Adan-Asim-ASE committed:
Applied migrations to DB and express session management along with few other issues

kamranarshad1 commented on .sequelizerc:
```diff
@@ -0,0 +1,5 @@
+
+module.exports = {
+  'config': path.resolve('config', 'config.js')
+}
```
Please add a new line at the end of file

kamranarshad1 commented on App.js:
```diff
@@ -0,0 +1,37 @@
+//     console.log( "Drop and re-sync db.");
+// }); 
+
+app.use(session({
+    secret: "thisismysecrctekey",
+    saveUninitialized: true,
+    cookie: { maxAge: 1000 * 60 * 60 * 24 },
+    resave: false
+    },));
```
We are developing an api so sessions don't make any sense. Please create authentication with jwt & passport

kamranarshad1 commented on config/config.js:
```diff
@@ -0,0 +1,25 @@
+    "password": process.env.DB_PASSWORD,
+    "database": process.env.DB_NAME,
+    "host": process.env.DB_HOST,
+    "dialect": "postgres"
```
dialects should also be coming from .env file

Adan-Asim-ASE committed:
Resolved minor highlighted issues

Adan-Asim-ASE committed:
Minor updations

Adan-Asim-ASE committed:
Replaced sessions with JWT and did other mentioned changes

Adan-Asim-ASE committed:
Minor updations

Adan-Asim-ASE committed:
Minor modifications

kamranarshad1 commented on App.js:
```diff
@@ -0,0 +1,31 @@
+
+// db.sequelize.sync({ force: true }).then(() => {          
+//     console.log( "Drop and re-sync db.");
+// }); 
```
please remove these

kamranarshad1 commented on App.js:
```diff
@@ -0,0 +1,31 @@
+const users = require('./routes/users.routes');
+const jwt = require("jsonwebtoken");
+
+require('dotenv').config();
```
Should be move to the top

kamranarshad1 commented on App.js:
```diff
@@ -0,0 +1,31 @@
+// }); 
+
+var corsOptions = {
+  origin: "http://localhost:3000"
```
Port should be as per env variable

kamranarshad1 commented on controllers/posts.controller.js:
```diff
@@ -0,0 +1,162 @@
+exports.createPost = (req, res) => {
+  if ((!req.body || !req.body.title || !req.body.content )) {
+    res.status(400).send({
+      message: "Post cannot be empty"
```
Please use `express-validator` to validate body and the message should be specific to that field not general

kamranarshad1 commented on controllers/posts.controller.js:
```diff
@@ -0,0 +1,162 @@
+
+  posts.create(post)
+    .then((post) => {
+      res.send(post);
```
Please add status as well with every response, for error, ok all type of http status as per scenario

kamranarshad1 replied:
Please don't use `send` use `json` method to return json responses

kamranarshad1 commented on controllers/posts.controller.js:
```diff
@@ -0,0 +1,162 @@
+      }
+      else {
+        res.status(404).send({
+          message: 'Cannot find any post with UserId=' + id
```
Dont reveal user id in error message. Why are you returning 404 for all posts request? This should just return empty array with status 200

kamranarshad1 commented on controllers/posts.controller.js:
```diff
@@ -0,0 +1,162 @@
+        res.status(404).send({
+          message: 'Cannot find any published post by User No: ' + id
+        });
+      }
```
Please fix this as per comments mentioned above

kamranarshad1 commented on controllers/posts.controller.js:
```diff
@@ -0,0 +1,162 @@
+      else {
+        res.status(404).send({
+          message: 'Cannot find any published post'
+        });
```
Please fix this

kamranarshad1 commented on controllers/posts.controller.js:
```diff
@@ -0,0 +1,162 @@
+  posts.findAll({ where: { published: true } })
+    .then((posts) => {
+      if (posts) {
+        res.send(posts);
```
status missing, use json instead

kamranarshad1 commented on controllers/posts.controller.js:
```diff
@@ -0,0 +1,162 @@
+
+
+exports.DeletePost = (req, res) => {
+  const pid = req.body.postId;
```
post id should be read from query params. First need to make sure if the current user is owner of the post else he return 401 error

Adan-Asim-ASE replied:
Need to make sure if the current user is owner of the post else he return 401 error, This check is done in posts.routes.js using a middleware defined in auth.js

kamranarshad1 commented on controllers/posts.controller.js:
```diff
@@ -0,0 +1,162 @@
+      }
+      else {
+        res.status(404).send({
+          message: 'Cannot find any post with id= ' + pid
```
Dont need to mention ids with error messages

kamranarshad1 commented on controllers/posts.controller.js:
```diff
@@ -0,0 +1,162 @@
+};
+
+exports.UpdatePost = (req, res) => {
+  const pid = req.body.postId;
```
post id should be read from query params. First need to make sure if the current user is owner of the post else he return 401 error

Adan-Asim-ASE replied:
First need to make sure if the current user is owner of the post else he return 401 error, This check is done in posts.routes.js using a middleware defined in auth.js

kamranarshad1 commented on controllers/users.controller.js:
```diff
@@ -0,0 +1,52 @@
+const users = db.users;
+
+exports.createUser = (req, res) => {
+  if ((!req.body || !req.body.name || !req.body.email || !req.body.password)) {
```
use express validator

kamranarshad1 commented on controllers/users.controller.js:
```diff
@@ -0,0 +1,52 @@
+  }
+  users.create(user)
+    .then((user) => {
+      const accessToken = jwt.sign({ user: user}, process.env.TOKEN_SECRET_KEY, { expiresIn: '1h' });
```
create separate method for getting creating jwt token

kamranarshad1 commented on controllers/users.controller.js:
```diff
@@ -0,0 +1,52 @@
+  const email = req.params.email;
+  const password = req.body.password;
+
+  users.findOne({ where: { email: email, password: password } })
```
Are you storing plain passwords in db? Passwords should be encrypted and then stored in the db, please use `argon2` for password encryption

kamranarshad1 commented on controllers/users.controller.js:
```diff
@@ -0,0 +1,52 @@
+  users.findOne({ where: { email: email, password: password } })
+    .then((user) => {
+      if (user != undefined) {
+        const accessToken = jwt.sign({ user: user}, process.env.TOKEN_SECRET_KEY, { expiresIn: '1h' });
```
Please move this to separate method

kamranarshad1 commented on controllers/users.controller.js:
```diff
@@ -0,0 +1,52 @@
+      }
+      else {
+        res.status(404).send({
+          message: "Cannot find user with email=" + email + " and password=" + password
```
Error message should be `invalid email or password`. Don't need to reveal email or password

kamranarshad1 commented on controllers/users.controller.js:
```diff
@@ -0,0 +1,52 @@
+    })
+    .catch((err) => {
+      res.status(500).send({
+        message: "Error retrieving user with email=" + email + " and password=" + password
```
Error message should be `something went wrong`

kamranarshad1 commented on migrations/20220706173321-create-users.js:
```diff
@@ -0,0 +1,33 @@
+        primaryKey: true,
+        type: Sequelize.INTEGER
+      },
+      name: {
+        type: Sequelize.STRING
+      },
+      password: {
+        type: Sequelize.STRING
+      },
+      email: {
+        type: Sequelize.STRING
+      },
```
Please add validations for not null

kamranarshad1 commented on migrations/20220706173321-create-users.js:
```diff
@@ -0,0 +1,33 @@
+        type: Sequelize.STRING
+      },
+      email: {
+        type: Sequelize.STRING
```
Please add unique constraint on the email

kamranarshad1 commented on migrations/20220706173546-create-posts.js:
```diff
@@ -0,0 +1,33 @@
+        type: Sequelize.INTEGER
+      },
+      title: {
+        type: Sequelize.STRING
```
Please add not null constraint for non-nullable attributes

kamranarshad1 commented on migrations/20220706173546-create-posts.js:
```diff
@@ -0,0 +1,33 @@
+      updatedAt: {
+        allowNull: false,
+        type: Sequelize.DATE
+      }
```
user id foreign key is missing

kamranarshad1 commented on models/posts.js:
```diff
@@ -0,0 +1,26 @@
+    }
+  }
+
+  posts.init({
+    title: DataTypes.STRING,
+    content: DataTypes.STRING,
+    published: DataTypes.BOOLEAN
+  }, 
```
validations missing in model

kamranarshad1 commented on models/users.js:
```diff
@@ -0,0 +1,22 @@
+module.exports = (sequelize, DataTypes) => {
+  class users extends Model {
+
+    static associate(models) {
```
Should mention association to posts so we can get posts by user

kamranarshad1 commented on routes/posts.routes.js:
```diff
@@ -0,0 +1,14 @@
+router.get("/me", auth, controller.findPostsByUserId);
+router.get("/me/published", auth, controller.findPublishedPostsByUserId);
+router.get("/me/drafted", auth, controller.findDraftedPostsByUserId);
+router.delete("/", auth, controller.DeletePost);
+router.post("/edit", auth, controller.UpdatePost);
```
Inconsistency in naming. Camelcasing should be used for method names

kamranarshad1 commented on controllers/users.controller.js:
```diff
@@ -0,0 +1,52 @@
+    });
+};
+
+exports.findUser = (req, res) => {
```
What's the purpose of this method?

Adan-Asim-ASE replied:
Its checks a whether a user exists, for login

kamranarshad1 replied:
If its a helper method, then it shouldn't be in controller

Adan-Asim-ASE committed:
App.js: removed extra code, used env variable in cops
Post.controller, User.controller: Used Expressed Validators, applied Argon encryption, changed res msgs and types status etc,
Post.model, User.model: Applied contraints and validations
Changed naming conventions to Camel Case

Adan-Asim-ASE committed:
Updation in post model, and create post controller

kamranarshad1 commented on App.js:
```diff
@@ -0,0 +1,29 @@
+const db = require("./models");
+const posts = require('./routes/posts.routes');
+const users = require('./routes/users.routes');
+
+
+
+
```
Remove extra lines

kamranarshad1 commented on controllers/posts.controller.js:
```diff
@@ -0,0 +1,158 @@
+        res.status(200).json(post);
+      }
+      else {
+        res.status(404).json({
+          message: 'Cannot find any post'
+        });
```
Only return 404 when there's specific resource lookup i.e. with id etc else just return response with 200

kamranarshad1 commented on controllers/posts.controller.js:
```diff
@@ -0,0 +1,158 @@
+        res.status(200).json(post);
+      }
+      else {
+        res.status(404).json({
+          message: 'Cannot find any published post'
+        });
```
Only return 404 when there's specific resource lookup i.e. with id etc else just return response with 200

kamranarshad1 commented on controllers/posts.controller.js:
```diff
@@ -0,0 +1,158 @@
+        res.status(200).json(posts);
+      }
+      else {
+        res.status(404).json({
+          message: 'Cannot find any published post'
+        });
```
Only return 404 when there's specific resource lookup i.e. with id etc else just return response with 200

kamranarshad1 commented on controllers/posts.controller.js:
```diff
@@ -0,0 +1,158 @@
+        res.status(200).json(post);
+      }
+      else {
+        res.status(404).json({
+          message: 'Cannot find any drafted post'
+        });
```
Only return 404 when there's specific resource lookup i.e. with id etc else just return response with 200

kamranarshad1 commented on controllers/users.controller.js:
```diff
@@ -0,0 +1,82 @@
+var argon2 = require('argon2');
+const users = db.users;
+
+generateToken = (user) => {
```
Move this to utils folder

kamranarshad1 commented on controllers/users.controller.js:
```diff
@@ -0,0 +1,82 @@
+  users.findOne({ where: { email: req.body.email } })
+    .then((user) => {
+      if (user != undefined) {
+        res.status(404).json({
```
status should be 400

kamranarshad1 commented on controllers/users.controller.js:
```diff
@@ -0,0 +1,82 @@
+              res.status(200).json({ accessToken });
+            }
+            else{
+              res.status(404).json({
```
status should be 400

kamranarshad1 commented on routes/users.routes.js:
```diff
@@ -0,0 +1,10 @@
+const userValidators = require("../validators/userValidators")
+
+
+router.post("/", userValidators.validateCreateUser,controller.createUser);
```
space missing before last middleware call

kamranarshad1 commented on routes/users.routes.js:
```diff
@@ -0,0 +1,10 @@
+
+
+router.post("/", userValidators.validateCreateUser,controller.createUser);
+router.post("/:email", userValidators.validateCheckUser,controller.checkUser);
```
space missing before last middleware call

kamranarshad1 replied:
Email & password both should be in form body. Route should be `/login`

kamranarshad1 replied:
`checkUser` is not a meaningful name. Please fix this

Adan-Asim-ASE committed:
App.js: Removed extra spaces
post.Controllers.js: Changed response codes
user.Controller.js: Changed response codes, moved token generation to Utils folder.
user.routes.js: Changed naming and spaces issues, moved email to req body

Adan-Asim-ASE committed:
post.controller.js: Changed return value in case of empty posts

kamranarshad1 commented on utils/generateToken.js:
```diff
@@ -0,0 +1,6 @@
+const jwt = require("jsonwebtoken");
+
+exports.generateToken = (user) => {
+    const accessToken = jwt.sign({ user: user }, process.env.TOKEN_SECRET_KEY, { expiresIn: '1h' });
```
This method should just receive payload & use that in signing

kamranarshad1 commented on utils/generateToken.js:
```diff
@@ -0,0 +1,6 @@
+exports.generateToken = (user) => {
+    const accessToken = jwt.sign({ user: user }, process.env.TOKEN_SECRET_KEY, { expiresIn: '1h' });
+    return accessToken;
+  }
```
Alignment issue

kamranarshad1 replied:
New line at the end of file

kamranarshad1 commented on controllers/posts.controller.js:
```diff
@@ -105,9 +99,7 @@ exports.findDraftedPostsByUserId = (req, res) => {
-        res.status(200).json({
-          message: 'Cannot find any drafted post'
-        });
+        res.status(200).json(post);
```
Both if & else cases are same so remove these conditions

Adan-Asim-ASE committed:
Configured ES lint and resolved issues highlighted by it

kamranarshad1 commented on validators/userValidators.js:
```diff
@@ -45,10 +45,10 @@ exports.validateUserLogin = [
     const errors = validationResult(req);
-    if (!errors.isEmpty())
-      return res.status(422).json({ errors: errors.array() });
+    if (!errors.isEmpty()) { return res.status(422).json({ errors: errors.array() }); }
```
Status should be 400

Adan-Asim-ASE committed:
Validators: Changed status code from 422 to 400

Adan-Asim-ASE committed:
Updated eslint
